ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs

```diff
@@ -377,6 +426,79 @@ extendVRNew cfg tx
                       , vrLastTicketNo = nextTicketNo
                       }
 
+-- | Attempt to validate and add as many new transactions to the mempool as
+-- possible, returning the last 'ValidationResult' and the remaining
+-- transactions which couldn't be added due to the mempool capacity being
+-- reached.
+validateNew :: forall m blk. (MonadSTM m, ApplyTx blk)
+            => MempoolEnv m blk
+            -> [GenTx blk]
+            -> ValidationResult blk
+            -> STM m (ValidationResult blk, [GenTx blk])
+validateNew mpEnv txs res =
+  let res' = res { vrInvalid = [] }
+  in case txs of
+    []                  -> return (res', [])
+    headTx:remainingTxs -> do
+      -- First, attempt to individually validate the first new transaction.
+      --
+      -- If this fails due to the mempool capacity being met, we should simply
+      -- 'retry' as it will be cheap due to the fact that we've done very
+      -- little work in this STM transaction.
+      --
+      -- If, however, this is successful, we should continue to validate all
+      -- of the other new transactions one-by-one. If the mempool capacity
+      -- would be reached at any step, we update the 'InternalState' with the
+      -- work that we've already done and return the last 'ValidationResult'
+      -- along with the remaining unvalidated transactions.
+      --
+      -- @intricate: I'm actually not sure if it makes sense to do this
+      --             specific validation of the head tx. 'go' looks like it's
+      --             sufficient for handling all of the transactions. If 'go'
+      --             fails on the first transaction (due to mempool cap), it
+      --             will return the remaining unvalidated transactions all the
+      --             way back up to 'implAddTxs' and, as a result, 'implAddTxs'
+      --             will recurse.
+      headTxValidationRes <- go [headTx] res'
+      case headTxValidationRes of
+        -- Mempool capacity hasn't been reached (no remaining unvalidated
+        -- transactions were returned).
+        (vr, []) -> do
+          -- Continue validating the remaining transactions.
+          (vr', unvalidatedTxs) <- go remainingTxs vr
```

Comment:
You are not validating these transactions one-by-one here.

So if we managed to add the first one, we try to add all the remaining ones in the same STM transaction (such that `retry`ing it redoes all the work), unless I'm not understanding it properly?

Reply:
I think the flow between `implAddTxs` and `validateNew` (and `go`) is getting a bit complicated. Maybe a comment explaining the flow, what is done within a single STM transaction, and what potentially has to be redone, could help. A diagram would be even nicer :grin:.

Additionally, this might sound strange, but making this a local function again, closer to `implAddTxs`, might be helpful (at least to me).

Reply:
> You are not validating these transactions one-by-one here.
> 
> So if we managed to add the first one, we try to add all the remaining ones in the same STM transaction (such that `retry`ing it redoes all the work), unless I'm not understanding it properly?

So yes, if we managed to add the first transaction, we then attempt to add the remaining transactions. `go` returns the last `ValidationResult` that it computed and a list of transactions that it couldn't continue validating due to the `MempoolCapacity` being reached. Therefore, if we reach `MempoolCapacity` while validating the remaining transactions, we update the mempool's `InternalState` with our progress thus far, return the list of unvalidated transactions all the way back up to `implAddTxs`, commit the `STM` transaction, and recursively call `implAddTxs` given that list of unvalidated transactions.

> I think the flow between `implAddTxs` and `validateNew` (and `go`) is getting a bit complicated. Maybe a comment explaining the flow, what is done within a single STM transaction, and what potentially has to be redone, could help. A diagram would be even nicer .
> 
> Additionally, this might sound strange, but making this a local function again, closer to `implAddTxs`, might be helpful (at least to me).

I agree that the flow is complicated. I think I would even find my explanation above a bit confusing had I not worked on the code myself. I'll see how I can simplify here and also add comments.

Reply:
Thanks for the explanation, I now understand it :slightly_smiling_face:
This clearly needs documentation :grin:

Reply:
Still working on this one. I've got some additional doc but it's incomplete so I haven't pushed it. 

I'm also a bit hesitant that the new names are more confusing than before: `validateNew` and `go` are now `validateNew` and `validateNewUntilMempoolFull`. May end up changing them again before this is ready to merge.

Reply:
Added some doc. If you think some information is missing, let me know! I believe it appropriately covers the flow between `implAddTxs`, `validateNew`, and `validateNewUntilMempoolFull`.